File: src/main/scala/stg/Detector.scala
package stg

import com.oracle.truffle.api.TruffleFile
import java.io.IOException
import java.nio.charset.Charset
import java.nio.{ ByteOrder, ByteBuffer }
import java.nio.file.StandardOpenOption
import scala.util.{ Try, Using }

class Detector extends TruffleFile.FileTypeDetector {
  @throws(classOf[IOException])
  override def findMimeType(file: TruffleFile): String = {
    val fileName = file.getName
    if (fileName == null) null
    else if (fileName.endsWith(Language.EXTENSION)) Language.MIME_TYPE
//    else if ((fileName.endsWith(Language.BYTECODE_EXTENSION) && (readMagicWord(file) == Language.BYTECODE_MAGIC_WORD))) Language.BYTECODE_MIME_TYPE
    else null
  }

  def readMagicWord(file: TruffleFile): Long = {
    val result : Try[Long] = Using (file.newInputStream(StandardOpenOption.READ)) { is => 
      val buffer = Array.fill[Byte](4)(0)
      if (is.read(buffer) != buffer.length) 0L
      else Integer.toUnsignedLong(ByteBuffer.wrap(buffer).order(ByteOrder.nativeOrder).getInt)
    }
    result.recover({
      case (_ : IOException) | (_ : SecurityException) => 0L
    }).get
  }

  @throws(classOf[IOException])
  override def findEncoding(file: TruffleFile): Charset = null
}

File: src/main/java/stg/Detector.java
package stg;

import com.oracle.truffle.api.TruffleFile;
import com.oracle.truffle.api.TruffleLanguage.Registration;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.Charset;
import java.nio.file.StandardOpenOption;

public class Detector implements TruffleFile.FileTypeDetector {
  @Override
  public String findMimeType(TruffleFile file) throws IOException {
    String name = file.getName();
    if (name != null && name.endsWith(Language.EXTENSION)) return Language.MIME_TYPE;

    //long magicWord = readMagicWord(file);
    //if (magicWord == Language.BYTECODE_MAGIC_WORD) return Language.BYTECODE_MIME_TYPE;
    return null;
  }

  @Override
  public Charset findEncoding(TruffleFile file) throws IOException {
    return null;
  }

  private static long readMagicWord(TruffleFile file) {
    try (InputStream is = file.newInputStream(StandardOpenOption.READ)) {
      byte[] buffer = new byte[4];
      if (is.read(buffer) != buffer.length) return 0;
      return Integer.toUnsignedLong(ByteBuffer.wrap(buffer).order(ByteOrder.nativeOrder()).getInt());
    } catch (IOException | SecurityException e) {
      return 0;
    }
  }
}
